src/Bicep.LangServer/Handlers/BicepForceModulesRestoreCommandHandler.cs

```diff
@@ -0,0 +1,110 @@
+    // It returns Restore (force) succeeded/failed message, which can be displayed approriately in IDE output window
+    public class BicepForceModulesRestoreCommandHandler : ExecuteTypedResponseCommandHandlerBase<string, string>
+    {
+        private readonly ICompilationManager compilationManager;
+        private readonly EmitterSettings emitterSettings;
+        private readonly IFeatureProvider features;
+        private readonly IFileResolver fileResolver;
+        private readonly IModuleDispatcher moduleDispatcher;
+        private readonly INamespaceProvider namespaceProvider;
+        private readonly IConfigurationManager configurationManager;
+
+        public BicepForceModulesRestoreCommandHandler(ICompilationManager compilationManager, ISerializer serializer, IFeatureProvider features, EmitterSettings emitterSettings, INamespaceProvider namespaceProvider, IFileResolver fileResolver, IModuleDispatcher moduleDispatcher, IConfigurationManager configurationManager)
+            : base(LangServerConstants.ForceModulesRestoreCommand, serializer)
+        {
+            this.compilationManager = compilationManager;
+            this.emitterSettings = emitterSettings;
+            this.features = features;
+            this.namespaceProvider = namespaceProvider;
+            this.fileResolver = fileResolver;
+            this.moduleDispatcher = moduleDispatcher;
+            this.configurationManager = configurationManager;
+        }
+
+        public override Task<string> Handle(string bicepFilePath, CancellationToken cancellationToken)
+        {
+            if (string.IsNullOrWhiteSpace(bicepFilePath))
+            {
+                throw new ArgumentException("Invalid input file path");
+            }
+
+            DocumentUri documentUri = DocumentUri.FromFileSystemPath(bicepFilePath);
+            Task<string> restoreOutput = GenerateForceModulesRestoreOutputMessage(bicepFilePath, documentUri);
+
+            return restoreOutput;
+        }
+
+        private async Task<string> GenerateForceModulesRestoreOutputMessage(string bicepFilePath, DocumentUri documentUri)
+        {
+            var fileUri = documentUri.ToUri();
+            RootConfiguration? configuration = null;
+
+            try
+            {
+                configuration = this.configurationManager.GetConfiguration(fileUri);
+            }
+            catch (ConfigurationException exception)
+            {
+                // Fail the restore if there's configuration errors.
+                return exception.Message;
+            }
+            Workspace workspace = new Workspace();
+            SourceFileGrouping sourceFileGrouping = SourceFileGroupingBuilder.Build(this.fileResolver, this.moduleDispatcher, workspace, fileUri, configuration);
+
+            // Ignore modules to restore logic, include all modules to be restored
+            var modulesToRestore = sourceFileGrouping.SourceFilesByModuleDeclaration
```

Comment:
This is the same logic that's in CLI. So if we put the flag in Build() like I mentioned in the other comment we could also simplify it here.

Reply:
Ok, so the idea would be to pass the `forceModulesRestore` bool value all the way down to `bicep/src/Bicep.Core/Workspaces/SourceFileGroupingBuilder.cs / private SourceFileGrouping Build()`... and down to `PopulateRecursive()` to centralize the logic for `ModuleRestoreStatus`? Down there we would force a value of `ModuleRestoreStatus.Unknown` if we see that `forceModulesRestore` is `true`?

I thought about this. It would be cleaner if `SourceFileGroupingBuilder.Build() / PopulateRecursive()` return all modules if `forceModulesRestore`. I was hesitant to go that route but you're probably right that the modules status / replacement logic should go somewhere centralized when we ask for a forced restore.